### PR TITLE
Fetch SPA bundle

### PR DESCRIPTION
**Pull request by oleh-momot**

### CODE REVIEW

File: registry/server/apps/routes/createApp.ts [hunk 1]
```diff
@@ -22,6 +24,33 @@ const validateRequestBeforeCreateApp = validateRequestFactory([{
 const createApp = async (req: Request, res: Response): Promise<void> => {
     const app = req.body;
 
+    if (!app.spaBundle && app.assetsDiscoveryUrl) {
+        let response;
+
+        try {
+            response = await axios.get(app.assetsDiscoveryUrl, {responseType: 'json'});
+        } catch (error) {
+            console.log('Caught an error while trying to fetch a manifest file:');
```
Comment (nightnei) [inline registry/server/apps/routes/createApp.ts:33]: add link(`app.assetsDiscoveryUrl`) to the end of the `console.log`

File: registry/server/apps/routes/createApp.ts [hunk 2]
```diff
@@ -22,6 +24,33 @@ const validateRequestBeforeCreateApp = validateRequestFactory([{
...
+            console.log(error);
```
Comment (nightnei) [inline registry/server/apps/routes/createApp.ts:34]: 1) why do you use `log` instead `error`?
2) you can provide the second argument to first `console.log`

**Review by nightnei — APPROVED** [review]

File: registry/client/src/apps/Edit.js [hunk 3]
```diff
@@ -1,74 +1,170 @@
...
+const selectHasAssetsDiscoveryUrl = (formData) => formData.assetsDiscoveryUrl && formData.assetsDiscoveryUrl.length !== 0;
+const selectWarnMessageDueToAssetsDiscoveryUrl = (formData) => {
+    if (!selectHasAssetsDiscoveryUrl(formData)) {
+        return '';
+    }
+
+    return `Do not need to specify SPA bundle, CSS bundle, dependencies because they would be fetched and set from assets discovery URL if they exist there`;
+};
+
+const validateAppCreation = (values) => {
```
Comment (StyleT) [inline registry/client/src/apps/Edit.js:31]: validateApp

Comment (StyleT) [reply registry/client/src/apps/Edit.js:31]: The code below is really complex. I don't really like the approach....

File: registry/server/common/services/json.ts [hunk 4]
```diff
@@ -13,23 +13,23 @@ const isJSON = (str: string): boolean => {
     return (/^[\],:{}\s]*$/).test(str);
 };
 
-const parse = (value: any): any => {
+const parseJSON = (value: any): any => {
```
Comment (StyleT) [inline registry/server/common/services/json.ts:16]: Why do you need to rename all these vars which unnecessarily cause a lot of side changes?

File: registry/server/common/services/preProcessResponse.ts [hunk 5]
```diff
@@ -1,25 +1,23 @@
 import _ from 'lodash/fp';
```
Comment (StyleT) [inline registry/server/common/services/preProcessResponse.ts:1]: Why do you need to rename all these vars which unnecessarily cause a lot of side changes?

File: registry/server/common/middlewares/validateRequest.ts [hunk 6]
```diff
@@ -17,19 +19,16 @@ interface ValidationConfig {
...
         await Promise.all(_.map(
             async ({schema, selector}) => schema.validateAsync(selector(req), {abortEarly: false}),
             validationConfig
         ));
         next();
-    } catch (e) {
-        res.status(422).send(preProcessErrorResponse(e));
+    } catch (error) {
+        console.error(`Caught an error while validating request data for '${req.url}':`, error);
+        res.status(422).send(preProcessErrorResponse(new Error(preProcessError(error))));
```
Comment (StyleT) [inline registry/server/common/middlewares/validateRequest.ts:31]: This line is super hard to understand....

File: registry/server/errorHandler/preProcessErrorResponse.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+const preProcessErrorResponse = (error: Error, errorInfo = {}) => {
```
Comment (StyleT) [inline registry/server/errorHandler/preProcessErrorResponse.ts:1]: I don't see a big value in this function... IMO it causes more issues then it solves

**Review by StyleT — CHANGES_REQUESTED** [review]
This PR contains a lot of unrelated to the main task changes and so it can't be merged

File: registry/server/common/middlewares/validateRequest.ts (hunk 6 above)
Comment (oleh-momot) [reply registry/server/common/middlewares/validateRequest.ts]: Removed

File: registry/server/errorHandler/preProcessErrorResponse.ts (hunk 7 above)
Comment (oleh-momot) [reply registry/server/errorHandler/preProcessErrorResponse.ts]: Removed

File: registry/server/common/services/json.ts (hunk 4 above)
Comment (oleh-momot) [reply registry/server/common/services/json.ts]: Removed

File: registry/server/common/services/preProcessResponse.ts (hunk 5 above)
Comment (oleh-momot) [reply registry/server/common/services/preProcessResponse.ts]: Removed

File: registry/client/src/apps/Edit.js (hunk 3 above)
Comment (oleh-momot) [reply registry/client/src/apps/Edit.js]: Renamed
Removed unnecessary code but I left necessary validation here for 'assetsDiscoveryUrl', 'spaBundle'

File: registry/client/src/apps/Edit.js [hunk 8]
```diff
@@ -12,36 +19,116 @@ import {
...
+                        return (
+                            <TextInput
+                                fullWidth
+                                resettable
+                                type="url"
+                                source="assetsDiscoveryUrl"
+                                disabled={isFetchingManifest}
+                                error={error}
+                                helperText={helperText}
+                                onBlur={(event) => {
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply registry/client/src/apps/Edit.js]: Removed

File: registry/client/src/apps/Edit.js [hunk 9]
```diff
@@ -12,36 +19,116 @@ import {
...
+                <TextInput
+                    fullWidth
+                    resettable
+                    type="url"
+                    source="spaBundle"
+                    validate={required()}
+                    disabled={isFetchingManifest}
+                    helperText={fetchingManifestHelperText}
+                />
+                <TextInput
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply registry/client/src/apps/Edit.js]: Added helper text for it

File: registry/client/src/apps/Edit.js [hunk 10]
```diff
@@ -12,36 +19,116 @@ import {
...
+                    source="configSelector"
+                    label="Shared props selector"
+                >
                     <AutocompleteArrayInput />
                 </ReferenceArrayInput>
             </FormTab>
             <FormTab label="Assets">
-                <TextInput source="spaBundle" validate={required()} fullWidth />
-                <TextInput source="cssBundle" fullWidth />
+                <TextInput
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply registry/client/src/apps/Edit.js]: Done

File: registry/server/apps/interfaces/index.ts [hunk 11]
```diff
@@ -35,5 +35,9 @@ export const partialAppSchema = Joi.object({
 export const appSchema = Joi.object({
     ...commonApp,
     name: appNameSchema.required(),
-    spaBundle: commonApp.spaBundle.required(),
+    spaBundle: Joi.when('assetsDiscoveryUrl', {
+        is: commonApp.assetsDiscoveryUrl.required(),
+        then: commonApp.spaBundle.optional(),
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply registry/server/apps/interfaces/index.ts]: I decided to leave optional here
It does not need to validate it, I guess

File: registry/server/apps/routes/createApp.ts [hunk 12]
```diff
@@ -22,6 +24,33 @@ const validateRequestBeforeCreateApp = validateRequestFactory([{
...
+        if (!processedManifest.spaBundle) {
+            res.status(422).send(`"spaBundle" must be specified in the manifest file from provided "assetsDiscoveryUrl" if it was not specified manually`);
+            return;
+        }
+
+        if (!app.cssBundle && processedManifest.cssBundle) {
+            app.cssBundle = processedManifest.cssBundle;
+        }
+
+        app.spaBundle = processedManifest.spaBundle;
```
[the remark this answers is not in the document]
Comment (oleh-momot) [reply registry/server/apps/routes/createApp.ts]: Done

File: registry/server/apps/routes/createApp.ts (hunk 2 above)
Comment (oleh-momot) [reply registry/server/apps/routes/createApp.ts]: Done

File: registry/server/apps/routes/createApp.ts (hunk 1 above)
Comment (oleh-momot) [reply registry/server/apps/routes/createApp.ts]: Done

**Review by StyleT — APPROVED** [review]

**Pull request merged**